src/pyhf/optimize/opt_tflow.py

```diff
@@ -1,76 +1,55 @@
 import tensorflow as tf
 
-log = logging.getLogger(__name__)
 
-
-class tflow_optimizer(object):
-    def __init__(self, tensorlib, **kwargs):
-        self.tb = tensorlib
-        self.relax = kwargs.get('relax', 0.1)
-        self.maxiter = kwargs.get('maxiter', 100000)
-        self.eps = kwargs.get('eps', 1e-4)
-
-    def unconstrained_bestfit(self, objective, data, pdf, init_pars, par_bounds):
-        # the graph
-        data = self.tb.astensor(data)
-        parlist = [self.tb.astensor([p]) for p in init_pars]
-
-        pars = self.tb.concatenate(parlist)
-        objective = objective(pars, data, pdf)
-        hessian = tf.hessians(objective, pars)[0]
-        gradient = tf.gradients(objective, pars)[0]
-        invhess = tf.linalg.inv(hessian)
-        update = tf.transpose(tf.matmul(invhess, tf.transpose(tf.stack([gradient]))))[0]
-
-        # run newton's method
-        best_fit = init_pars
-        for i in range(self.maxiter):
-            up = self.tb.session.run(update, feed_dict={pars: best_fit})
-            best_fit = best_fit - self.relax * up
-            if np.abs(np.max(up)) < self.eps:
-                break
-
-        return best_fit.tolist()
-
-    def constrained_bestfit(
-        self, objective, constrained_mu, data, pdf, init_pars, par_bounds
-    ):
-        # the graph
-        data = self.tb.astensor(data)
-
-        nuis_pars = [
-            self.tb.astensor([p])
-            for i, p in enumerate(init_pars)
-            if i != pdf.config.poi_index
-        ]
-        poi_par = self.tb.astensor([constrained_mu])
-
-        nuis_cat = self.tb.concatenate(nuis_pars)
-        pars = self.tb.concatenate(
-            [
-                nuis_cat[: pdf.config.poi_index],
-                poi_par,
-                nuis_cat[pdf.config.poi_index :],
-            ]
+def eval_func(op, argop, dataop, d):
+    def func(p):
+        tensorlib, _ = get_backend()
+        xx = tensorlib.tolist(p) if isinstance(p, tf.Tensor) else p
+        yy = tensorlib.tolist(d) if isinstance(d, tf.Tensor) else d
+        value = tensorlib.session.run(op, feed_dict={argop: xx, dataop: yy})
+        return value
+
+    return func
+
+
+class tflow_optimizer(AutoDiffOptimizerMixin):
+    def __init__(self, *args, **kargs):
+        pass
+
+    def setup_unconstrained(self, objective, m, d, init_pars, par_bounds):
+        pars = tf.placeholder(tf.float32, (m.config.npars,))
+        data = tf.placeholder(tf.float32, (m.config.nmaindata + m.config.nauxdata,))
+        nll = objective(pars, data, m)
+        nllgrad = tf.identity(tf.gradients(nll, pars)[0])
+        func = eval_func([nll, nllgrad], pars, data, d)
+        return func, init_pars, par_bounds
+
+    def setup_constrained(self, objective, poival, m, d, init_pars, par_bounds):
+        tensorlib, _ = get_backend()
+        data = tf.placeholder(tf.float32, (m.config.nmaindata + m.config.nauxdata,))
+        idx = default_backend.astensor(range(m.config.npars), dtype='int')
```

Comment:
probably should support `tensorlib.range`

Reply:
PRs welcome :)

Reply:
#636 is filed for now.